crates/resolver/src/name_server/name_server.rs

```diff
@@ -143,21 +149,40 @@ impl<C: DnsHandle, P: ConnectionProvider<Conn = C>> NameServer<C, P> {
                         ResponseCode::ServFail => {
                             let note = "Nameserver responded with SERVFAIL";
                             debug!("{}", note);
-                            return Err(ProtoError::from(note));
+                            let error_kind = ResolveErrorKind::Proto(ProtoError::from(note));
+                            return Err(ResolveError::from(error_kind));
                         }
                         ResponseCode::NXDomain => {
                             let note = "Nameserver responded with NXDomain";
                             debug!("{}", note);
-                            return Err(ProtoError::from(note));
+
+                            let mut response = response;
+                            let query =
+                                response.take_queries().drain(..).next().unwrap_or_default();
+                            let error_kind = ResolveErrorKind::NoRecordsFound {
+                                query,
+                                valid_until: None,
+                                response_code: ResponseCode::NXDomain,
+                            };
+                            return Err(ResolveError::from(error_kind));
                         }
                         ResponseCode::NoError
                             if response.answers().is_empty()
                                 && response.additionals().is_empty()
                                 && response.name_servers().is_empty() =>
```

Comment:
we only care about name_servers in the case where we are looking up specific record types, NS, PTR, NAPTR. We may need a specific check for this?